File: Level 1/가장 많이 받은 선물/solution.js
function solution(friends, gifts) {
  const friendMap = {};
  const nextMonthReceiveMap = {};

  friends.forEach((friend) => {
    friendMap[friend] = { send: [], receive: [] };
    nextMonthReceiveMap[friend] = 0;
  });

  gifts.forEach((gift) => {
    const [sender, receiver] = gift.split(" ");
    friendMap[sender].send.push(receiver);
    friendMap[receiver].receive.push(sender);
  });

  for (let i = 0; i < friends.length; i++) {
    for (let j = i + 1; j < friends.length; j++) {
      const A = friends[i];
      const B = friends[j];

      const AToBGift = friendMap[A].send.filter(
        (friend) => friend === B
      ).length;
      const BToAGift = friendMap[B].send.filter(
        (friend) => friend === A
      ).length;

      if (AToBGift === BToAGift) {
        const AGiftScore =
          friendMap[A].send.length - friendMap[A].receive.length;
        const BGiftScore =
          friendMap[B].send.length - friendMap[B].receive.length;
        if (AGiftScore > BGiftScore) {
          nextMonthReceiveMap[A] += 1;
        }
        if (BGiftScore > AGiftScore) {
          nextMonthReceiveMap[B] += 1;
        }
      } else {
        if (AToBGift > BToAGift) {
          nextMonthReceiveMap[A] += 1;
        }
        if (BToAGift > AToBGift) {
          nextMonthReceiveMap[B] += 1;
        }
      }
    }
  }

  return Math.max(...Object.values(nextMonthReceiveMap));
}

File: Level 1/가장 많이 받은 선물/solution.py
def solution(friends, gifts):
    gift_map = {friend: {"receive": [], "send": []} for friend in friends}

    for gift in gifts:
        sender, receiver = gift.split(" ")
        gift_map[sender]["send"].append(receiver)
        gift_map[receiver]["receive"].append(sender)

    next_month_receive_map = {friend: 0 for friend in friends}

    for i in range(len(friends)):
        for j in range(i + 1, len(friends)):
            A = friends[i]
            B = friends[j]

            A_to_B_gift = gift_map[A]["send"].count(B)
            B_to_A_gift = gift_map[B]["send"].count(A)

            if A_to_B_gift == B_to_A_gift:
                A_gift_score = len(gift_map[A]["send"]) - len(gift_map[A]["receive"])
                B_gift_score = len(gift_map[B]["send"]) - len(gift_map[B]["receive"])
                if A_gift_score > B_gift_score:
                    next_month_receive_map[A] += 1
                elif B_gift_score > A_gift_score:
                    next_month_receive_map[B] += 1
            else:
                if A_to_B_gift > B_to_A_gift:
                    next_month_receive_map[A] += 1
                elif B_to_A_gift > A_to_B_gift:
                    next_month_receive_map[B] += 1

    return max(next_month_receive_map.values())
